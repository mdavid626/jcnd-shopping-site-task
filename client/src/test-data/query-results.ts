export const productsQueryResult = {
  data: {
    products: {
      nodes: [
        {
          id: '63ea6485732ff5cf8b364431',
          priceInCents: 9699,
          name: 'Broccoli',
          inStock: 466,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364432',
          priceInCents: 3434,
          name: 'Broccolini',
          inStock: 223,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364433',
          priceInCents: 3692,
          name: 'Brussels Sprouts',
          inStock: 12,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364434',
          priceInCents: 8839,
          name: 'Cabbage',
          inStock: 200,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364435',
          priceInCents: 6828,
          name: 'Carrots',
          inStock: 151,
          __typename: 'Product',
        },
      ],
      pageInfo: { page: 1, numberOfPages: 15, __typename: 'PageInfo' },
      __typename: 'ProductsQueryResult',
    },
  },
};

export const placeOrderQueryResult1 = {
  data: { placeOrder: '63ede5e031f3fc4012a0951b' },
};
