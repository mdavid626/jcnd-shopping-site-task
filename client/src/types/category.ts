export type Category = 'vegetables' | 'fruits' | 'cheese';
