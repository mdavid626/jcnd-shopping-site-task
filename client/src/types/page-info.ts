export type PageInfo = {
  page: number;
  numberOfPages: number;
};
