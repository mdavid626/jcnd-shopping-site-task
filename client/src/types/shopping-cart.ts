export type ShoppingCartItem = {
  productId: string;
  amount: number;
};
