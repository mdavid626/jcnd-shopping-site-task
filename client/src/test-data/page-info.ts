import { PageInfo } from '../types/page-info';

export const testPageInfo1: PageInfo = {
  page: 1,
  numberOfPages: 5,
};
