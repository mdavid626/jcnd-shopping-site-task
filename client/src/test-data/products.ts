import { Product, ProductsQueryResult } from '../types/products';

export const testProduct1: Product = {
  id: '63ea6485732ff5cf8b36442c',
  name: 'Artichoke',
  price: 46.36,
  inStock: 5,
};

export const testProduct2: Product = {
  id: '63ea6485732ff5cf8b36442d',
  name: 'Asparagus',
  price: 25.36,
  inStock: 3,
};

export const testProductQueryResult1: ProductsQueryResult = {
  products: {
    nodes: [testProduct1, testProduct2],
    pageInfo: {
      page: 1,
      numberOfPages: 2,
    },
  },
};
