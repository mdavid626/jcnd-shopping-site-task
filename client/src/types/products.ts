import { PageInfo } from './page-info';

export type ProductCategory = 'VEGETABLES' | 'FRUITS' | 'CHEESE';

export type Product = {
  id: string;
  price: number;
  name: string;
  inStock: number;
};

export type ProductsQueryResult = {
  products: { nodes: Product[]; pageInfo: PageInfo };
};
