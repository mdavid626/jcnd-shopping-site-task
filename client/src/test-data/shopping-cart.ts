import { ShoppingCartItem } from '../types/shopping-cart';
import { testProduct1, testProduct2 } from './products';

export const testShoppingCartItem1: ShoppingCartItem = {
  product: testProduct1,
  amount: 3,
};

export const testShoppingCartItem2: ShoppingCartItem = {
  product: testProduct2,
  amount: 36,
};

export const testShoppingCart1: ShoppingCartItem[] = [
  testShoppingCartItem1,
  testShoppingCartItem2,
];
