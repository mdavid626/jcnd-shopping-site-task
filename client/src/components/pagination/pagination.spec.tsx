import { cleanup, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { testPageInfo1 } from '../../test-data/page-info';
import { PageInfo } from '../../types/page-info';
import Pagination from './pagination';

describe('pagination', () => {
  afterEach(cleanup);

  it('should render', () => {
    const { asFragment } = render(
      <Pagination
        pageInfo={testPageInfo1}
        goNext={jest.fn()}
        goPrevious={jest.fn()}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should add custom className when passed', () => {
    render(
      <Pagination
        pageInfo={testPageInfo1}
        goNext={jest.fn()}
        goPrevious={jest.fn()}
        className="myClassName"
      />
    );
    expect(screen.getByTestId('Pagination')).toHaveClass('myClassName');
  });

  it('should be disabled when no pageInfo', () => {
    render(<Pagination goNext={jest.fn()} goPrevious={jest.fn()} />);
    expect(screen.getByTitle('Go to previous page')).not.toHaveClass(
      'Pagination-arrow--enabled'
    );
    expect(screen.getByTitle('Go to next page')).not.toHaveClass(
      'Pagination-arrow--enabled'
    );
  });

  it('should disable go to previous page when on first page', () => {
    const pageInfo: PageInfo = {
      page: 0,
      numberOfPages: 5,
    };
    render(
      <Pagination
        pageInfo={pageInfo}
        goNext={jest.fn()}
        goPrevious={jest.fn()}
      />
    );
    expect(screen.getByTitle('Go to previous page')).not.toHaveClass(
      'Pagination-arrow--enabled'
    );
    expect(screen.getByTitle('Go to next page')).toHaveClass(
      'Pagination-arrow--enabled'
    );
  });

  it('should disable go to next page when on last page', () => {
    const pageInfo: PageInfo = {
      page: 4,
      numberOfPages: 5,
    };
    render(
      <Pagination
        pageInfo={pageInfo}
        goNext={jest.fn()}
        goPrevious={jest.fn()}
      />
    );
    expect(screen.getByTitle('Go to previous page')).toHaveClass(
      'Pagination-arrow--enabled'
    );
    expect(screen.getByTitle('Go to next page')).not.toHaveClass(
      'Pagination-arrow--enabled'
    );
  });

  it('should go previous when go previous page clicked', async () => {
    const goPrevious = jest.fn();
    render(
      <Pagination
        pageInfo={testPageInfo1}
        goNext={jest.fn()}
        goPrevious={goPrevious}
      />
    );
    await userEvent.click(screen.getByTitle('Go to previous page'));
    expect(goPrevious).toHaveBeenCalledWith();
  });

  it('should go next when go next page clicked', async () => {
    const goNext = jest.fn();
    render(
      <Pagination
        pageInfo={testPageInfo1}
        goNext={goNext}
        goPrevious={jest.fn()}
      />
    );
    await userEvent.click(screen.getByTitle('Go to next page'));
    expect(goNext).toHaveBeenCalledWith();
  });
});
