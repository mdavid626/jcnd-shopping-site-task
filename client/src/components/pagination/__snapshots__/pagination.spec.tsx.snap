// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pagination should render 1`] = `
<DocumentFragment>
  <div
    class="Pagination"
    data-testid="Pagination"
  >
    <div
      class="Pagination-arrow Pagination-arrow--enabled"
      title="Go to previous page"
    >
      «
    </div>
    <div
      class="Pagination-arrow Pagination-arrow--enabled"
      title="Go to next page"
    >
      »
    </div>
  </div>
</DocumentFragment>
`;
